### PR TITLE
imp: rm app upgrade interface from IBCModule and use type assertions for app callback routing

### PR DESCRIPTION
**Pull request by damiannolan**
## Description

- Removes `UpgradableModule` from `IBCModule` interface
- Uses type assertions to cast `cbs` to `UpgradeableModule` for app callback routing in core

closes: #1614

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/core/05-port/types/module.go [hunk 1]
```diff
@@ -2,8 +2,8 @@ package types
 
 import (
 	sdk "github.com/cosmos/cosmos-sdk/types"
-
 	capabilitytypes "github.com/cosmos/ibc-go/modules/capability/types"
```
Comment (colin-axner) [inline modules/core/05-port/types/module.go:5]: interesting, this doesn't fail the linter? (for unformatted imports?)

**Review by colin-axner — APPROVED** [review]
LGTM, but would be very lovely with test cases :)

File: modules/core/05-port/types/module.go (hunk 1 above)
Comment (damiannolan) [reply modules/core/05-port/types/module.go:5]: I expect it probably does! Seems like there's a big queue atm for runners. Was about to push a `make lint-fix` commit which changes these lines

**Comment by damiannolan** [comment]
> LGTM, but would be very lovely with test cases :)

Indeed, I thought about tests but I think the mock app is implementing all the upgrade callbacks. I'd rather not introduce a dependency on some app that doesn't implement the interface *yet*. 
We'd probably need to add another mock app struct somewhere, what do you think?

**Comment by damiannolan** [comment]
Think we need to cast in the app callbacks as well

File: modules/core/keeper/msg_server.go [hunk 2]
```diff
@@ -794,12 +800,18 @@ func (k Keeper) ChannelUpgradeTry(goCtx context.Context, msg *channeltypes.MsgCh
...
-	cbs, ok := k.Router.GetRoute(module)
+	app, ok := k.Router.GetRoute(module)
 	if !ok {
 		ctx.Logger().Error("channel upgrade try failed", "port-id", msg.PortId, "error", errorsmod.Wrapf(porttypes.ErrInvalidRoute, "route not found to module: %s", module))
 		return nil, errorsmod.Wrapf(porttypes.ErrInvalidRoute, "route not found to module: %s", module)
 	}
 
+	cbs, ok := app.(porttypes.UpgradableModule)
+	if !ok {
+		ctx.Logger().Error("channel upgrade try failed", "port-id", msg.PortId, "error", errorsmod.Wrapf(porttypes.ErrInvalidRoute, "upgrade route not found to module: %s", module))
```
[the remark this answers is not in the document]
Comment (damiannolan) [reply modules/core/keeper/msg_server.go:811]: Aye! They do indeed. You're referring to the error within the log and the return statement, correct?
Looks like its a similar pattern used across many of the funcs in this file (not just the upgrade handlers). 

I can address these ones here - create error and reuse for both log and return, and follow up with a PR to main for the rest.

File: modules/apps/27-interchain-accounts/controller/ibc_middleware.go [hunk 3]
```diff
@@ -233,6 +233,11 @@ func (im IBCMiddleware) OnTimeoutPacket(
 
 // OnChanUpgradeInit implements the IBCModule interface
 func (im IBCMiddleware) OnChanUpgradeInit(ctx sdk.Context, portID, channelID string, order channeltypes.Order, connectionHops []string, version string) (string, error) {
+	cbs, ok := im.app.(porttypes.UpgradableModule)
+	if !ok {
+		return "", errorsmod.Wrapf(porttypes.ErrInvalidRoute, "upgrade route not found to module in application callstack")
```
[the remark this answers is not in the document]
Comment (damiannolan) [reply modules/apps/27-interchain-accounts/controller/ibc_middleware.go:238]: good catch

File: modules/core/keeper/msg_server.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (damiannolan) [reply modules/core/keeper/msg_server.go:811]: edit: actually these might be all better left to a single PR. I can open an issue

**Comment by colin-axner** [comment]
> Indeed, I thought about tests but I think the mock app is implementing all the upgrade callbacks. I'd rather not introduce a dependency on some app that doesn't implement the interface yet.
We'd probably need to add another mock app struct somewhere, what do you think?

I would test by creating a router which points to a mock ibc app which doesn't implement the upgrade module interface. It might be possible to rework the mock package so that it has three types: 
- IBCModule
- UpgradeModule
- MockModule (contains both IBCModule and UpgradeModule)

**Pull request merged**